src/decorators.py

```diff
@@ -1,167 +1,220 @@
-
 import fnmatch
+from collections import defaultdict
 
 import botconfig
-import src.settings as var
 from oyoyo.parse import parse_nick
+from src import settings as var
 from src import logger
 
 adminlog = logger(None)
 
-def generate(fdict, permissions=True, **kwargs):
-    """Generates a decorator generator.  Always use this"""
-    def cmd(*s, raw_nick=False, admin_only=False, owner_only=False, chan=True, pm=False,
-                game=False, join=False, none=False, playing=False, roles=(), hookid=-1):
-        def dec(f):
-            def innerf(*args):
-                largs = list(args)
-                rawnick = largs[1]
-                if not permissions:
-                    return f(*largs)
-                if len(largs) > 1 and largs[1]:
-                    nick, _, _, cloak = parse_nick(largs[1])
-
-                    if cloak is None:
-                        cloak = ""
-                else:
-                    nick = ""
-                    cloak = ""
-                if not raw_nick and len(largs) > 1 and largs[1]:
-                    largs[1] = nick
-                if nick == "<console>":
-                    return f(*largs) # special case; no questions
-                if not pm and largs[2] == nick: # PM command
-                    return
-                if not chan and largs[2] != nick: # channel command
-                    return
-                if largs[2].startswith("#") and largs[2] != botconfig.CHANNEL and not admin_only and not owner_only:
-                    if "" in s:
-                        return # Don't have empty commands triggering in other channels
-                    allowed = False
-                    for cmdname in s:
-                        if cmdname in botconfig.ALLOWED_ALT_CHANNELS_COMMANDS:
-                            allowed = True
-                            break
-                    if not allowed:
-                        return
-                if nick in var.USERS.keys() and var.USERS[nick]["account"] != "*":
-                    acc = var.USERS[nick]["account"]
+COMMANDS = defaultdict(list)
+HOOKS = defaultdict(list)
+
+class cmd:
+    def __init__(self, *cmds, raw_nick=False, admin_only=False, owner_only=False,
+                 chan=True, pm=False, join=False, none=False, game=False, playing=False, roles=()):
```

Comment:
Consider condensing the join, none, game params into a single phase param, takes in a tuple/list/whatever like phase=("join","day","night") (could have "game" as an alias for both "day" and "night")

(I believe @nyuszika7h is the one that originally came up with this idea, just stating it in github comments for posterity's sake)